### ISSUE TITLE
S3 Media Storage does not truncate existing file when opened for write

### ISSUE THREAD

**Issue by gwharton** [magento/magento2#36789]
### Preconditions and environment

- Magento 2.4.5-p1

### Steps to reproduce

Download test module Test.zip https://lubefindercom-my.sharepoint.com/:u:/g/personal/graham_lubefinder_com/ESLHS90FdOxDrENcAYkimNQB1SGlTQHbvnt370MXMZajYA?e=gjDUF5

Install Test module onto vanilla Magento
```
> php bin/magento setup:upgrade

> php bin/magento gw:test:writeinitialfile
Writing Test File
Done
> cat pub/media/testfile.txt
This is a test string written by the gw:test:writeinitialfile command
> php bin/magento gw:test:overwritefile
Overwriting Test File
Done
> cat pub/media/testfile.txt
Overwritten by gw:test:overwritefile
>
```

Now configure S3 storage for media and repeat the test

### Expected result

```
> php bin/magento gw:test:writeinitialfile
Writing Test File
Done
> download and view contents of pub/media/testfile.txt from S3
This is a test string written by the gw:test:writeinitialfile command
> php bin/magento gw:test:overwritefile
Overwriting Test File
Done
> download and view contents of pub/media/testfile.txt from S3
Overwritten by gw:test:overwritefile
>
```

### Actual result

```
> php bin/magento gw:test:writeinitialfile
Writing Test File
Done
> download and view contents of pub/media/testfile.txt from S3
This is a test string written by the gw:test:writeinitialfile command
> php bin/magento gw:test:overwritefile
Overwriting Test File
Done
> download and view contents of pub/media/testfile.txt from S3
Overwritten by gw:test:overwritefile gw:test:writeinitialfile command
>
```

Note that the file was not truncated before being written to. The behaviour should be the same as running the test on a file based media storage.

### Release note

S3 Media Storage truncates existing file contents when overwriting existing files.

### Triage and priority

- [X] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-Hotel** [comment]
Hello @gwharton,

Thanks for the report and contribution!

We have tried to reproduce the issue in the latest development branch i.e. 2.4-develop and the issue is not reproducible for us via the module provided by you in the main description.

It is working as expected in local as well as S3. Please find attached the txt file while writing and overwriting with the commands:

While Writing:
[testfile-write.txt](https://github.com/magento/magento2/files/10533524/testfile-write.txt)

While Overwrite:
[testfile-overwrite.txt](https://github.com/magento/magento2/files/10533527/testfile-overwrite.txt)

Can you please recheck the reproduction steps and let us know if anything has been missed?

Thanks

**Comment by gwharton** [comment]
The overwritten file in your post does show the problem. It contains

Overwritten by gw:test:overwritefile **gw:test:writeinitialfile command**

The bold part is from the original write. The overwrite command writes "Overwritten by gw:test:overwritefile" only.

**Comment by gwharton** [comment]
The first command writes "This is a test string written by the gw:test:writeinitialfile command"

The second command overwrites with "Overwritten by gw:test:overwritefile"

The fact that the file contains the string "Overwritten by gw:test:overwritefile gw:test:writeinitialfile command" suggests that the fault was reproduced.

**Comment by engcom-Hotel** [comment]
Hello @gwharton,

Thanks for the reply!

Yes, you are correct, this was missed by me. I am confirming this issue for further processing.

Thanks

**Comment by Nuranto** [comment]
Maybe linked to https://github.com/magento/magento2/issues/37584

**Comment by theuargb** [comment]
@magento I am working on this

**Comment by Nuranto** [comment]
Hello @theuargb,

Any news on this ? Can I help ?

**Comment by Nuranto** [comment]
@magento I am working on this

**Comment by Nuranto** [comment]
Oh, it seems it's already fixed in `2.4-develop` branch !

**Comment by engcom-Hotel** [comment]
Hello @Nuranto,

We have rechecked the issue, and you are correct. It got fixed. 

Hence we are closing this issue.

Thanks
